platforms/nuttx/src/bootloader/microchip/mpfs/main.c

```diff
@@ -491,32 +486,72 @@ flash_func_write_word(uintptr_t address, uint32_t word)
 
 #if defined(CONFIG_MTD_M25P) || defined(CONFIG_MMCSD)
 
-	// Write a single page every time we got a full one
+	static uintptr_t end_address = 0;
```

Comment:
I'm having a hard time understanding why these static variables are needed, also, where do these point if one transfer to disk fails and another attempt is needed ? There is no reset mechanism for these as far as I can see ?

Reply:
it tries to write to mass media in blocks, but the function is called for every word. So it is collecting the data into DRAM, and keeping track of the progress in the static variables.

end_address is specifically needed in the end of flashing, when less than a full block has been written to the DRAM. Then you need some way to know what is the last word that was written, to be able to write the last incomplete block of data.

Reply:
Yes exactly, it keeps track of the progress for one process. Good optimization for flash write access but if the process fails / is interrupted, the static variables still point to where the previous process left off. I don't know if this is an issue. Then again, the status of the flash media is also changed and needs erase.

Reply:
When a new flashing starts, the static variables are updated again. The write of the new image is always sequential from 0 ->, the end_address is set first time on the second word written, the first_unwritten is set in the end of every write. Both are used only when writing the last word (which is writing the 0 word again with value different than 0xffffffff). So these two don't need reset logic.